### PR TITLE
[WIP] Gif Renderer

### PR DESCRIPTION
**Pull request by Dominaezzz**
Add component for https://github.com/JetBrains/compose-jb/issues/153

### CODE REVIEW

**Comment by olonho** [comment]
Do not think that this is GIF-only renderer, think it provide support for wider set of animations supported by Skia. So better rename it.

File: components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import androidx.compose.ui.graphics.painter.Painter
+import org.jetbrains.skija.Bitmap
+import org.jetbrains.skija.Codec
+import org.jetbrains.skija.Data
+import java.net.URL
+
+@Composable
+internal actual fun rememberGifRendererImpl(url: String): State<Painter> {
+    val codec = remember(url) {
+        println("Gif renderer for $url")
```
Comment (olonho) [inline components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt:16]: Avoid debug printouts.

File: components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        val bytes = URL(url).readBytes()
```
Comment (olonho) [inline components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt:17]: Use IO dispatcher to fetching data from network.

File: components/GifRenderer/common/src/commonMain/kotlin/org/jetbrains/compose/gifrenderer/GifRenderer.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package org.jetbrains.compose.gifrenderer
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.State
+import androidx.compose.ui.graphics.painter.Painter
+
+@Composable
+fun rememberGifRenderer(url: String): State<Painter> {
```
Comment (igordmn) [inline components/GifRenderer/common/src/commonMain/kotlin/org/jetbrains/compose/gifrenderer/GifRenderer.kt:8]: I think we need to provide a way to load gif from any `InputStream` and from any non-composable function (that can be called from `Dispatchers.IO`)

```
fun loadAnimation(inputStream: InputStream): Animation

// load synchronously from resources (see https://android-review.googlesource.com/c/platform/frameworks/support/+/1576433/2/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt#71)
@Composable
fun animationResource(resourcePath: String)

@Composable
fun Animation(animation: Animation) {

}

interface Animation {
  val frames: Int
  fun painterFor(frame: Int): Painter
}
```

P.S. `InputStream` is jvm-only, so it should be part of the `desktopMain`. 
Maybe we can implement something using https://github.com/Kotlin/kotlinx-io, but it is experimental, so our function should be experimental too.

File: components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt (hunk 2 above)
Comment (igordmn) [reply components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt:17]: I think we can provide an universal function to load from any InputStream (see above). But we shouldn't decide for the users how they want to load an animation - synchronously or asynchronously.

So I suggest not to introduce the Composable API for loading from URL. In core Compose we also didn't introduce API for asynchronous loading (to be specific, we introduced it in the past, but removed [later](https://android-review.googlesource.com/c/platform/frameworks/support/+/1561297))

File: components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    val frameIndex by transition.animateValue(
+        initialValue = 0,
+        targetValue = codec.frameCount - 1,
+        Int.VectorConverter,
+        animationSpec = infiniteRepeatable(
+            animation = keyframes {
+                durationMillis = 0
+                for ((index, frame) in codec.framesInfo.withIndex()) {
+                    index at durationMillis
+                    durationMillis += frame.duration
```
Comment (YeungKC) [inline components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt:41]: > If the delay time of a frame is less than 20ms, mainstream browsers will set 100ms by default.
https://github.com/flutter/flutter/issues/29130#issuecomment-897732165

if use https://mixin-images.zeromesh.net/rtCKv3Mng6KY7ydK9KhKJfD7NL1QkqSsBnMd-u4sS74XId7cjLBiWPVvIrJucKOhyp0tg3YZhqTM3t5hAEoAtRU will found, only first one frame have a duration, other frame’s duration is zero.

**Comment by igordmn** [comment]
Closed as there is no activity. We still plan to make it as a separate component in https://github.com/JetBrains/compose-jb/tree/master/components, when we find time to do it.

**Pull request closed without merging**